### PR TITLE
Enable BatchNorm to use global mean and variane during training

### PR DESCRIPTION
**Pull request by qingqing01**
- Enable BatchNorm to use global mean and variane during training
- Add unit testing
- Usage:
   set `use_global_stats=True` in `fluid.layers.batch_norm` arguments.

### CODE REVIEW

File: paddle/fluid/operators/batch_norm_op.cc [hunk 1]
```diff
@@ -419,38 +447,61 @@ class BatchNormGradKernel<platform::CPUDeviceContext, T>
                                           : x_dims[x_dims.size() - 1]);
     const int sample_size = x->numel() / N / C;
 
-    ConstEigenVectorArrayMap<T> scale_arr(scale->data<T>(), C);
-    ConstEigenVectorArrayMap<T> mean_arr(saved_mean->data<T>(), C);
-    ConstEigenVectorArrayMap<T> inv_var_arr(saved_inv_variance->data<T>(), C);
+    auto *d_x = ctx.Output<Tensor>(framework::GradVarName("X"));
+    d_x->mutable_data<T>(ctx.GetPlace());
+    if ((N * sample_size) == 1) {
+      framework::TensorCopySync(*d_y, ctx.GetPlace(), d_x);
```
Comment (panyx0718) [inline paddle/fluid/operators/batch_norm_op.cc]: why is this sync copy?

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -2349,6 +2350,12 @@ def batch_norm(input,
         moving_variance_name(string, Default None): The name of the moving_variance which store the global Variance.
         do_model_average_for_mean_and_var(bool, Default False): Do model average for mean and variance or not.
         fuse_with_relu (bool): if True, this OP performs relu after batch norm.
+        use_global_stats(bool, Default False): Whether to use global mean and
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2366]: what does it mean by "global"? All BN use the same mean and variance? Or the BN use the moving mean and variable?

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -2349,6 +2350,12 @@ def batch_norm(input,
...
+            variance. In inference or test mode, set use_global_stats to true
+            or is_test to true, and the behavior is equivalent.
+            In train mode, when setting use_global_stats True, the global mean
+            and variance are also used during train time,
+            the BN acts as scaling and shiffting.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: What does this mean: `the BN acts as scaling and shiffting.`

BN always acts as scaling and shifting? How is it related to global stats?

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -2381,9 +2388,15 @@ def batch_norm(input,
         shape=param_shape,
         dtype=dtype,
         default_initializer=Constant(1.0))
+    # setting stop_gradient=True to reduce computation
+    if use_global_stats and helper.param_attr.learning_rate == 0.:
+        scale.stop_gradient = True
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2405]: learning_rate = 0 is equal to stop_gradient=True?

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -2381,9 +2388,15 @@ def batch_norm(input,
...
 
     bias = helper.create_parameter(
         attr=helper.bias_attr, shape=param_shape, dtype=dtype, is_bias=True)
+    # setting stop_gradient=True to reduce computation
+    if use_global_stats and helper.bias_attr.learning_rate == 0.:
+        scale.stop_gradient = True
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2411]: same as above

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2366]: Doc the math somewhere?

File: paddle/fluid/operators/batch_norm_op.cc (hunk 1 above)
Comment (qingqing01) [reply paddle/fluid/operators/batch_norm_op.cc]: `TensorCopy` is also ok. I change to `TensorCopy`. Thanks!

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/fluid/layers/nn.py:2366]: Yeah,  use the same mean and variance during training period. It is usually happens in fine-tuning, the `global_stats` are the moving mean and variance got by in the pre-trained model.

sometimes, global, running, moving are same meaning in batch norm.

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (qingqing01) [reply python/paddle/fluid/layers/nn.py]:
```
y = scale * ((x - global_mean) / sqrt(global_variance))  + bias
```
Here, I also called above formula  scaling and shifting.

I remove this line and add math.

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/layers/nn.py:2405]: Yeah, lr=0 will lead to the fixed parameter. But not set `stop_gradient=True`, the op still calculates the gradient. If set ture, not calculate.

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/fluid/layers/nn.py:2366]: Done.

**Review by typhoonzero — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
LG API
